eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Utxow.hs

```diff
@@ -0,0 +1,216 @@
+-- Here we define the transtion function, using reusable tests.
+-- The tests are very generic and reusable, but the transition
+-- function is very specific to the Babbage Era.
+
+-- | A very specialized transitionRule function for the Babbage Era.
+conwayUtxowTransition ::
+  forall era.
+  ( AlonzoEraTx era
+  , ExtendedUTxO era
+  , EraUTxO era
+  , ScriptsNeeded era ~ AlonzoScriptsNeeded era
+  , Script era ~ AlonzoScript era
+  , ConwayEraTxBody era
+  , Signable (DSIGN (EraCrypto era)) (Hash (HASH (EraCrypto era)) EraIndependentTxBody)
+  , -- Allow UTXOW to call UTXO
+    Embed (EraRule "UTXO" era) (ConwayUTXOW era)
+  , Environment (EraRule "UTXO" era) ~ UtxoEnv era
+  , Signal (EraRule "UTXO" era) ~ Tx era
+  , State (EraRule "UTXO" era) ~ UTxOState era
+  ) =>
+  TransitionRule (ConwayUTXOW era)
+conwayUtxowTransition = do
+  (TRC (UtxoEnv slot pp stakepools genDelegs, u, tx)) <- judgmentContext
+
+  {-  (utxo,_,_,_ ) := utxoSt  -}
+  {-  txb := txbody tx  -}
+  {-  txw := txwits tx  -}
+  {-  witsKeyHashes := { hashKey vk | vk ∈ dom(txwitsVKey txw) }  -}
+  let utxo = utxosUtxo u
+      txBody = tx ^. bodyTxL
+      witsKeyHashes = witsFromTxWitnesses @era tx
+      hashScriptMap = txscripts utxo tx
+      inputs = (txBody ^. referenceInputsTxBodyL) `Set.union` (txBody ^. inputsTxBodyL)
+
+  -- check scripts
+  {- neededHashes := {h | ( , h) ∈ scriptsNeeded utxo txb} -}
+  {- neededHashes − dom(refScripts tx utxo) = dom(txwitscripts txw) -}
+  let scriptsNeeded = getScriptsNeeded utxo txBody
+      scriptHashesNeeded = getScriptsHashesNeeded scriptsNeeded
+  {- ∀s ∈ (txscripts txw utxo neededHashes ) ∩ Scriptph1 , validateScript s tx -}
+  -- CHANGED In BABBAGE txscripts depends on UTxO
+  runTest $ validateFailedBabbageScripts tx utxo scriptHashesNeeded
+
+  {- neededHashes − dom(refScripts tx utxo) = dom(txwitscripts txw) -}
+  let sReceived = Map.keysSet $ tx ^. witsTxL . scriptTxWitsL
+      sRefs = Map.keysSet $ refScripts inputs utxo
+  runTest $ babbageMissingScripts pp scriptHashesNeeded sRefs sReceived
+
+  {-  inputHashes ⊆  dom(txdats txw) ⊆  allowed -}
+  runTest $ missingRequiredDatums hashScriptMap utxo tx
+
+  {-  dom (txrdmrs tx) = { rdptr txb sp | (sp, h) ∈ scriptsNeeded utxo tx,
+                           h ↦ s ∈ txscripts txw, s ∈ Scriptph2}     -}
+  runTest $ hasExactSetOfRedeemers utxo tx scriptsNeeded
+
+  -- check VKey witnesses
+  -- let txbodyHash = hashAnnotated @(Crypto era) txbody
+  {-  ∀ (vk ↦ σ) ∈ (txwitsVKey txw), V_vk⟦ txbodyHash ⟧_σ                -}
+  runTestOnSignal $ Shelley.validateVerifiedWits tx
+
+  {-  witsVKeyNeeded utxo tx ⊆ witsKeyHashes                   -}
+  let needed = conwayWitsVKeyNeeded utxo (tx ^. bodyTxL)
+  runTest $ validateNeededWitnesses @era witsKeyHashes needed
+
+  -- check metadata hash
+  {-   adh := txADhash txb;  ad := auxiliaryData tx                      -}
+  {-  ((adh = ◇) ∧ (ad= ◇)) ∨ (adh = hashAD ad)                          -}
```

Comment:
I'm surprised that `fourmolu` didn't convert all the comments to `-- ` style. I remember in a previous PR that it did that for me. :slightly_smiling_face:

Reply:
I think it does it only for haddock `{-| ... -}` -> `-- |`

Reply:
Oh I see :+1: